server/config/component_config.go

```diff
@@ -0,0 +1,398 @@
+	config, err := c.getComponentCfg(component, componentID)
+	if err != nil {
+		status = &configpb.Status{Code: configpb.Status_UNKNOWN, Message: "encode error"}
+		return latestVersion, "", status
+	}
+
+	return latestVersion, config, status
+}
+
+func (c *ComponentsConfig) getLatestVersion(component, componentID string) *configpb.Version {
+	v := &configpb.Version{
+		Global: c.GlobalCfgs[component].GetVersion(),
+		Local:  c.LocalCfgs[componentID].GetVersion().GetLocal(),
+	}
+	return v
+}
+
+func (c *ComponentsConfig) getComponentCfg(component, componentID string) (string, error) {
+	config := c.LocalCfgs[componentID].GetConfigs()
+	updateEntries := make(map[string]*EntryValue)
+	// apply the global change to updateEntries
+	if _, ok := c.GlobalCfgs[component]; ok {
+		globalUpdateEntries := c.GlobalCfgs[component].GetUpdateEntries()
+		for k, v := range globalUpdateEntries {
+			updateEntries[k] = v
+		}
+	}
+
+	// apply the local change to updateEntries
+	for k1, v1 := range c.LocalCfgs[componentID].GetUpdateEntries() {
+		if v, ok := updateEntries[k1]; ok {
+			// apply conflict
+			if v1.Version.GetGlobal() == v.Version.GetGlobal() {
+				updateEntries[k1] = v1
+			}
+		} else {
+			updateEntries[k1] = v1
+		}
+	}
+
+	for k, v := range updateEntries {
+		configName := strings.Split(k, ".")
+		update(config, configName, v.Value)
+	}
+
+	return encodeConfigs(config)
+}
+
+// Update ...
+// We denote rv for the request version, pv for the version stored in PD.
+// 1. rv >= pv -> pv = rv + 1, update config, return pv
+// 2. rv < pv -> return pv
+func (c *ComponentsConfig) Update(kind *configpb.ConfigKind, version *configpb.Version, entries []*configpb.ConfigEntry) (*configpb.Version, *configpb.Status) {
+	c.Lock()
+	defer c.Unlock()
+
+	global := kind.GetGlobal()
+	if global != nil {
+		component := global.GetComponent()
+		// if the global config of the component is existed.
+		if globalCfg, ok := c.GlobalCfgs[component]; ok {
+			globalLatestVersion := globalCfg.GetVersion()
+			if globalLatestVersion > version.GetGlobal() {
+				return &configpb.Version{Global: globalLatestVersion, Local: version.GetLocal()},
+					&configpb.Status{
+						Code:    configpb.Status_STALE_VERSION,
+						Message: "version is wrong",
+					}
+			} else if globalLatestVersion < version.GetGlobal() {
+				return &configpb.Version{Global: globalLatestVersion, Local: version.GetLocal()},
+					&configpb.Status{
+						Code:    configpb.Status_UNKNOWN,
+						Message: "version is wrong",
+					}
+			}
+
+			newGlobalVersion := version.GetGlobal() + 1
+			for _, entry := range entries {
+				globalCfg.updateEntry(entry, &configpb.Version{Global: newGlobalVersion, Local: 0})
+			}
+			globalCfg.Version = newGlobalVersion
+			for _, lc := range c.LocalCfgs {
+				lc.Version = &configpb.Version{Global: newGlobalVersion, Local: 0}
+			}
+		} else {
+			globalCfg := NewGlobalConfig(entries, &configpb.Version{Global: 1, Local: 0})
+			c.GlobalCfgs[component] = globalCfg
```

Comment:
Maybe we should create global config for each component after PD is started, since we only support few limited components (TiDB & TiKV) right now. And reject update request for unknown component.